### ISSUE TITLE
Quick Forms: Add asset checkboxes for each location

### ISSUE THREAD

**Issue by aislinnpearson** [Rothamsted-Ecoinformatics/farm_rothamsted#289]
Something I overlooked when specifying #260 is that if you select a location you _must_ have at least one asset selected which is associated with that location. 

For example the below will assign an asset to Great Field but not Dell Bones:

![image](https://user-images.githubusercontent.com/16349781/190864762-ff5fc19a-3e0c-44d1-9b7f-59da98c690c7.png)

We had this problem the other day where the entered the correct location but _didn't_ select the asset associated with that location, and then when we reviewed the log realised it was missing one of the assets. 

It is an easy enough mistake if you are doing a log that covers a lot of fields at once so I was wondering if we could include a warning/ error message to say that:

1. the user has selected a location but not the associated asset (prompting to select the asset or delete the location from the list). 
2. Appears this error/ warning would come up at the start of the quick form before moving to the next tab
3. Ideally ensures you wouldn't be able to submit the form without having at least one asset associated with every location selected.

**Comment by paul121** [comment]
> Ideally ensures you wouldn't be able to submit the form without having at least one asset associated with every location selected.

Oh interesting. One challenge I see here is that from the long list of checkboxes it might be challenging to identify which assets are associated with which location. Because many of your assets have a name that is prefixed with the location this may be less of an issue for Rothamsted, but that could be an issue for generalizing it.

Another potential issue here relates to #293 where we are displaying assets across the entire hierarchy as asset options. I guess this is largely a technical issue as it makes it harder to validate the user's selection... we just have to check against the entire hierarchy, not direct parents.

@aislinnpearson I want to clarify one thing, right now these selected "Locations" are not being saved to the log. Since they are not being saved I think this is less of an issue maybe. But if you do want them saved to the log, then I agree that we should enforce that an asset is selected for each location...

**Comment by aislinnpearson** [comment]
On the call on Tuesday we agreed that the location(s) would also be saved to the logs along with the assets (also true for previous implementations of the quick forms, although locations weren't previously selected). @paul121 also suggested that 

1. we could display the location name with the assets underneath, to make it clear which locations don't have assets selected
2. if an asset has two locations (e.g. a single plant asset drilled over two fields) it would appear as selected in both locations if you select one asset

Paul also wanted to make it clear that this isn't the ideal implementation and this added complexity is more likely to be prone to errors, but it's one of the few parts of the user interface which is designed to meet human preference. Particular care therefore needs to be taken when this part of the quick forms is included in any other issues such as #287

**Comment by paul121** [comment]
Done! A few notes...

@aislinnpearson I'd like you to specifically test adding locations, checking checkboxes, and then removing locations that have assets checked. Especially if you have 3 locations, and you check assets for the third, but remove the second location. This is not something I originally considered. My assumption is that the form will "forget" those assets checked for the 3rd location. This is not ideal, but unfortunately I think it is just a limitation of this approach.

@aislinnpearson I'll let you close this when you've tested & approved..

> if an asset has two locations (e.g. a single plant asset drilled over two fields) it would appear as selected in both locations if you select one asset

I did not add this. I believe we decided it was OK to make the user select an asses multiple times if it appeared for multiple locations.

Generally, I hope this won't happen often. If it does happen often, I think it will be because when you add a location, we actually search for assets in sub-locations of that location as well. If a user adds both a location its sub-locations into the "Locations" area then there will likely be duplicate asset checkboxes. If this becomes an issue I would prefer we re-visit the design before adding more complexity.

Here is a screenshot, I added a fieldset to make it more clear how location & assets work together:

![Screenshot from 2022-10-24 16-47-40](https://user-images.githubusercontent.com/3116995/197651553-473fde29-38de-462a-b7ab-1a65f688638b.png)

**Comment by aislinnpearson** [comment]
Hey @paul121 

I tried testing this one too but was limited by the tractor/ equipment groupings again - also, I think I need the different location types for fields to be enabled. One for going through this evening - maybe we can chat about how this works in terms of #301 if @richardostler joins us. It's quite a lot to add every time we want to test a release...

**Comment by aislinnpearson** [comment]
HI @paul121,

Had another go at testing this today and you are correct - if I add a location and then select an asset, but later delete the location _without_ selecting the tick box, then there is an error: the form submits _with_ the asset even though the location has been deleted. So good catch there - that would have been a major problem, and the inverse of the issue we originally had. :) 

How do you want to deal with this? It sounds to me like we can't get it to happen automatically, but could we add logic to say that the form can't be submitted if there is no location associated with an asset? 

> I did not add this. I believe we decided it was OK to make the user select an asses multiple times if it appeared for multiple locations.

I'm trying to think this through in my head. Internally we agreed that if an asset covers two fields, both fields need to be linked to the asset. This is definitely a tricky one because it would mean the log saves the correct asset in all instances, but only one of the two locations. I think this is rare - what actually happens is that we have fields which are now merged under Chris' management and will _always_ now be managed as a single field, so where it has happened I've created a new field which is a combination of the two previous fields. It's an edge case, but maybe we should open an new issue for it and come back to it in February/ March next year once the field net replacement is done?

**Comment by paul121** [comment]
> If I add a location and then select an asset, but later delete the location without selecting the tick box, then there is an error: the form submits with the asset even though the location has been deleted.

Aha, this is a slightly different issue than what I was envisioning but I think it also solves what I had described. Sorry if that wasn't clear but I think we've gotten to the bottom of it :-)

I changed the "event" that triggers the ajax from `autocompleteclose` to `change`. This will trigger an ajax load when items are removed. A side effect with this change is that you must "click" outside of the text box after making/changing selections in order to trigger an ajax load. With `autocompleteclose` this would load immediately after selecting an item, but I guess it doesn't account for removing items.

**Comment by paul121** [comment]
@aislinnpearson this is ready to test on the demo site!

**Comment by aislinnpearson** [comment]
Thanks @paul. I just tested it now - there is a slightly strange behaviour in that you have to press 'enter' to get the form to do what you want (or click elsewhere, like selecting a log category to get the form to update). However, it submits correctly which is great. I think I am happy to close this issue so we can get on with entering data, and put that in as a minor issue to discuss next week. :)
